File: app/src/main/java/com/fernandosierra/door2door/presentation/base/BasePresenter.java
package com.fernandosierra.door2door.presentation.base;

import android.support.annotation.NonNull;

import java.lang.ref.WeakReference;

public abstract class BasePresenter<V extends BaseView> {
    private final WeakReference<V> viewRef;

    public BasePresenter(@NonNull V view) {
        this.viewRef = new WeakReference<>(view);
    }

    public void view(@NonNull SecureViewCallback<V> secureViewCallback) {
        V view = viewRef.get();
        if (view != null) {
            secureViewCallback.execute(view);
        }
    }

    public interface SecureViewCallback<V extends BaseView> {
        void execute(@NonNull V view);
    }
}

File: app/src/main/java/com/fernandosierra/door2door/presentation/base/BasePresenter.kt
package com.fernandosierra.door2door.presentation.base

import java.lang.ref.WeakReference

abstract class BasePresenter<out V : BaseView>(view: V) {
    private val viewRef = WeakReference(view)

    fun view(block: (V) -> Unit) {
        val view = viewRef.get()
        if (view != null) {
            block.invoke(view)
        }
    }
}